### ISSUE TITLE
Kernel Panic after Restart of WireGuard

### ISSUE THREAD

**Issue by gshaffer1** [theonemcdonald/pfSense-pkg-WireGuard#82]
Installed pfSense-pkg-WireGuard-0.1.1_1.txz (over v0.0.9). After status showed tunnels were back up, tried pinging several systems on the other side of a site-to-site tunnel. Pinging the IPv6 addresses was successful, but pinging the IPv4 addresses timed out. From the WireGuard Status page, "Restarted: the service, box became unresponsive and rebooted.

From /var/system.log (more output is contained in the attached file):

May 14 14:52:42 egis kernel: Fatal trap 12: page fault while in kernel mode
May 14 14:52:42 egis kernel: cpuid = 0;
May 14 14:52:42 egis kernel: apic id = 00
May 14 14:52:42 egis kernel: fault virtual address      = 0x10
May 14 14:52:42 egis kernel: fault code         = supervisor read data, page not present
May 14 14:52:42 egis kernel: instruction pointer        = 0x20:0xffffffff8450817d
May 14 14:52:42 egis kernel: stack pointer              = 0x28:0xfffffe00b1872340
May 14 14:52:42 egis kernel: frame pointer              = 0x28:0xfffffe00b18723f0
May 14 14:52:42 egis kernel: code segment               = base 0x0, limit 0xfffff, type 0x1b
May 14 14:52:42 egis kernel:                    = DPL 0, pres 1, long 1, def32 0, gran 1
May 14 14:52:42 egis kernel: processor eflags   = interrupt enabled,
May 14 14:52:42 egis kernel: resume, IOPL = 0
May 14 14:52:42 egis kernel: current process            = 97644 (bgpd)
May 14 14:52:42 egis kernel: trap number                = 12
May 14 14:55:06 egis syslogd: kernel boot file is /boot/kernel/kernel
May 14 14:55:06 egis kernel: panic: page fault
May 14 14:55:06 egis kernel: cpuid = 0
May 14 14:55:06 egis kernel: time = 1621029162
May 14 14:55:06 egis kernel: KDB: enter: panic

Box is running FRR/BGP.

After the reboot, I noticed the following alert:

"PHP errors

 PHP ERROR: Type: 64, File: /usr/local/pkg/wireguard/wg_service.inc, Line: 361, Message: Cannot redeclare print_message() (previously declared in /usr/local/etc/rc.reload_wireguard:36) @ 2021-05-14 14:54:59"

I let the system settle for about ten minutes and restarted WireGuard without any issues.

What other info do you need to look at this?

[system.log](https://github.com/theonemcdonald/pfSense-pkg-WireGuard/files/6481605/system.log)

**Comment by theonemcdonald** [comment]
Interesting. What versions of the kmod and tools are you running?

**Comment by gshaffer1** [comment]
> Interesting. What versions of the kmod and tools are you running?
> 
> Can you confirm if rc.reload_wireguard is present on disk? It shouldn't be, as it's functionality has been replaced by the daemon. Also make sure the only early shell command in your config is 'service wireguardd start'
> 
> You might also try running pkg remove and pkg add again , I'm really perplexed why rc.reload_wireguard is still sticking around

Running:
- wireguard-kmod-0.0.20210503
- wireguard-tools-lite-1.0.20210424

I did not reload them with v0.1.1_1 as they didn't appear to be different from what was loaded with v0.0.9.

rc.reload_wireguard is present on my box in /usr/local/etc

Here are the bits from my config.xml:

                <earlyshellcmd>/usr/local/bin/php-cgi -f /usr/local/bin/apply_patches.php</earlyshellcmd>
                <earlyshellcmd>/usr/local/etc/rc.reload_wireguard</earlyshellcmd>
                <earlyshellcmd>service wireguardd start</earlyshellcmd>

and

                <shellcmdsettings>
                        <config>
                                <cmd>/usr/local/etc/rc.reload_wireguard</cmd>
                                <cmdtype>earlyshellcmd</cmdtype>
                                <description><![CDATA[WireGuard earlyshellcmd (DO NOT EDIT/DELETE!)]]></description>
                        </config>
                        <config>
                                <cmd>service wireguardd start</cmd>
                                <cmdtype>earlyshellcmd</cmdtype>
                                <description><![CDATA[WireGuard earlyshellcmd (DO NOT EDIT/DELETE!)]]></description>
                        </config>
                </shellcmdsettings>

I will wait on removing and reinstalling the package for a little while in case you need anything else.

**Comment by theonemcdonald** [comment]
We've been able to partially replicate this internally. Remove any references to rc.reload_wireguard in your config XML and on disk. Also, if you're side loading, it's best to first pkg remove, then pkg add. This ensures all old files are removed. Everything on disk is ephemeral, just make sure keep configuration is ticked in settings and you can safely nuke the package and reinstall fresh

**Comment by gshaffer1** [comment]
> We've been able to partially replicate this internally. Remove any references to rc.reload_wireguard in your config XML and on disk. Also, if you're side loading, it's best to first pkg remove, then pkg add. This ensures all old files are removed. Everything on disk outside of the config XML is ephemeral, just make sure keep configuration is ticked in settings and you can safely nuke the package and reinstall fresh

Did a "pkg remove". Looking at the config.xml, the early shell command bits were still there, even the new wireguard service (exactly the same as reported above). Removed both, deleted rc.reload_wireguard and /tmp/config.cache and rebooted. Reinstalled just the pfsense-pkg-WireGuard-0.1.1_1. Services showed the Wireguard as "not started", status page showed the mod had not been loaded and both tunnels were greyed out. Editing a tunnels and saving it without changing anything eventually brought up both tunnels. Like yesterday, I could ping the IPv6 addresses of several systems on the other side of the tunnel, but not the IPv4 address. Rebooted the box, after it settled, I could still not ping the IPv4 addresses even though the routes show up in the routing table. After restarting FRR/BGP I was able to ping both IPv4 and IPv6 addresses.

Good news, no kernel panics, but there appears to be some kind of timing issue with FRR/BGP.
